### PR TITLE
Fix noteplan 3 extension

### LINKED ISSUE #2007: [NotePlan 3] ...
# Extension – [NotePlan 3](https://github.com/raycast/extensions/tree/cd83a235882f060ff70c4d6e8d7255758f790b18/extensions/noteplan-3/)
Author: @shakedlokits

## Description
Command "Search Notes" doesn't search anything
## Steps To Reproduce

1.Run in Raycast Noteplan3 Search Notes
2. type anything
3. see No result

### Raycast version
Version: 1.36.0

### PR DESCRIPTION
**Pull request by lukerandall**
## Description

Fixes NotePlan 3 when using an extension other than `.txt` for your notes. Fixes #2007 

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

### CODE REVIEW

**Comment by danielsequeira** [comment]
@shakedlokits would you mind do a review here?

**Comment by lukerandall** [comment]
@danielsequeira What's the expected process here if @shakedlokits doesn't respond?

**Comment by mattisssa** [comment]
> @danielsequeira What's the expected process here if @shakedlokits doesn't respond?

We will take a look at the changes ourselves and merge them 🙃 
I guess 10 days is enough to do so.

File: extensions/noteplan-3/src/lib/note-utilities.ts [hunk 1]
```diff
@@ -25,8 +30,16 @@ export interface Note {
   content: string;
 }
 
+const getPreferences = () => {
+  return getPreferenceValues<Preferences>();
+};
```
Comment (mattisssa) [inline extensions/noteplan-3/src/lib/note-utilities.ts:35]: I guess you could simply use 
```suggestion
export const { preferences } = getPreferenceValues<Preferences>()
```
here, or do you need it to be a closure?

File: extensions/noteplan-3/src/lib/note-utilities.ts [hunk 2]
```diff
@@ -25,8 +30,16 @@ export interface Note {
...
+
+export const notesExtension = () => {
+  return getPreferences().extension;
+};
```
Comment (mattisssa) [inline extensions/noteplan-3/src/lib/note-utilities.ts:39]:
```suggestion
```

File: extensions/noteplan-3/src/lib/note-utilities.ts [hunk 3]
```diff
@@ -25,8 +30,16 @@ export interface Note {
...
+
 export const listNotes = (): NoteEntry[] => {
-  const paths = find(`${NOTE_PLAN_URI}/**/*.txt`, { absolute: true });
+  const paths = find(`${NOTE_PLAN_URI}/Notes/**/*.${notesExtension()}`, { absolute: true });
```
Comment (mattisssa) [inline extensions/noteplan-3/src/lib/note-utilities.ts:42]: Why have you changed the path by adding `/Notes/` here? I'm not familiar with the noteplan but wouldn't it break some cases?

```suggestion
  const paths = find(`${NOTE_PLAN_URI}/Notes/**/*.${preferences.extension}`, { absolute: true });
```

File: extensions/noteplan-3/src/today.tsx [hunk 4]
```diff
@@ -1,11 +1,11 @@
-import { NoteType } from "./lib/note-utilities";
+import { NoteType, notesExtension } from "./lib/note-utilities";
 import { NoteDetail } from "./lib/components";
 import { format } from "date-fns";
 
 export default () => (
   <NoteDetail
     entry={{
-      relativePath: `Calendar/${format(new Date(), "yyyyMMdd")}.txt`,
+      relativePath: `Calendar/${format(new Date(), "yyyyMMdd")}.${notesExtension()}`,
```
Comment (mattisssa) [inline extensions/noteplan-3/src/today.tsx:8]:
```suggestion
      relativePath: `Calendar/${format(new Date(), "yyyyMMdd")}.${preferences.extension}`,
```

File: extensions/noteplan-3/src/lib/note-utilities.ts (hunk 3 above)
Comment (lukerandall) [reply extensions/noteplan-3/src/lib/note-utilities.ts:42]: To be honest I'm surprised it worked before; I assume the original author doesn't have any backups (these are triggered manually) to mess things up. NotePlan's app support directory structure:

CleanShot 2022-08-01 at 15 16 26@2x

File: extensions/noteplan-3/src/today.tsx [hunk 5]
```diff
@@ -1,11 +1,11 @@
-import { NoteType } from "./lib/note-utilities";
+import { NoteType, notesExtension } from "./lib/note-utilities";
```
Comment (lukerandall) [inline extensions/noteplan-3/src/today.tsx:1]:
```suggestion
import { NoteType, preferences } from "./lib/note-utilities";
```

**Comment by shakedlokits** [comment]
Hi there! Sorry I was on vacation.
@danielsequeira, do you know if the preferences backfills the defaults for users who already had the extension installed?

File: extensions/noteplan-3/src/lib/note-utilities.ts [hunk 6]
```diff
@@ -14,6 +15,10 @@ export enum NoteType {
   Project = "project",
 }
 
+interface Preferences {
```
Comment (shakedlokits) [inline extensions/noteplan-3/src/lib/note-utilities.ts:18]: Let's extract the preferences to another file and only expose a getter function, i.e
`export const getPreferences = () => getPreferenceValues<Preferences>();`

File: extensions/noteplan-3/src/today.tsx [hunk 7]
```diff
@@ -1,11 +1,11 @@
-import { NoteType } from "./lib/note-utilities";
+import { NoteType, preferences } from "./lib/note-utilities";
 import { NoteDetail } from "./lib/components";
 import { format } from "date-fns";
 
 export default () => (
   <NoteDetail
     entry={{
-      relativePath: `Calendar/${format(new Date(), "yyyyMMdd")}.txt`,
+      relativePath: `Calendar/${format(new Date(), "yyyyMMdd")}.${preferences.extension}`,
```
Comment (shakedlokits) [inline extensions/noteplan-3/src/today.tsx:8]: Do we know if these preferences are being backfilled with the default for users who already had the extension?

File: extensions/noteplan-3/package.json [hunk 8]
```diff
@@ -28,6 +29,16 @@
...
     }
   ],
+  "preferences": [
+    {
+      "name": "extension",
+      "type": "textfield",
+      "required": false,
+      "title": "Note extension",
+      "default": "txt",
+      "description": "The extension you use for notes"
```
Comment (shakedlokits) [inline extensions/noteplan-3/package.json:39]: Let's try something more explicit and have similar naming for the title and name, like
```suggestion
      "name": "fileExtension",
      "type": "textfield",
      "required": false,
      "title": "File Extension",
      "default": "txt",
      "description": "The extension you use for notes"
```

**Review by shakedlokits — COMMENTED** [review]
Sorry for the long haul, looks great! Added some minor comments👌🏽

**Comment by lukerandall** [comment]
> Hi there! Sorry I was on vacation.

No worries! Hope it was a good one!

> @mattisssa, do you know if the preferences backfills the defaults for users who already had the extension installed?

I just tested this after renaming the preference and it defaults to the `default` - ie `txt` without any configuration required.

File: extensions/noteplan-3/src/lib/note-utilities.ts (hunk 6 above)
Comment (lukerandall) [reply extensions/noteplan-3/src/lib/note-utilities.ts]: Yeah, that makes sense.

File: extensions/noteplan-3/package.json (hunk 8 above)
Comment (lukerandall) [reply extensions/noteplan-3/package.json]: 👍

File: extensions/noteplan-3/src/lib/preferences.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+import { getPreferenceValues } from "@raycast/api";
+
+interface Preferences {
+  fileExtension: string;
+}
+
+export const getPreferences = getPreferenceValues<Preferences>();
```
Comment (shakedlokits) [inline extensions/noteplan-3/src/lib/preferences.ts:7]: Can we make this a function? So we don't initialize on import

**Comment by shakedlokits** [comment]
> > Hi there! Sorry I was on vacation.
> 
> 
> 
> No worries! Hope it was a good one!
> 
> 
> 
> > @mattisssa, do you know if the preferences backfills the defaults for users who already had the extension installed?
> 
> 
> 
> I just tested this after renaming the preference and it defaults to the `default` - ie `txt` without any configuration required.
> 
> 
> 
> 

Great,thanks! I added a really small comment on initializing on import but other than LGTM, damn fine work!👌🏽

**Review by shakedlokits — APPROVED** [review]
amaze-gif

**Comment by lukerandall** [comment]
@shakedlokits I've made the change suggested. Thanks for your feedback!

**Comment by mattisssa** [comment]
> Hi there! Sorry I was on vacation.
> @mattisssa, do you know if the preferences backfills the defaults for users who already had the extension installed?

yes, sure, it will be fulfilled with default values.

**Comment by mattisssa** [comment]
By the way @shakedlokits @lukerandall, I don't have the noteplan 3 installed and still have extension working, _(it's actually empty for every screen)_. Would be nice in the future to show some toast/empty screen that an end-user doesn't have the app installed 😉

otherwise, I just see this error message any time I perform an action.

CleanShot 2022-08-02 at 12 42 42@2x

**Pull request merged**